### PR TITLE
[SPARK-9818] Re-enable Docker tests for JDBC data source

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch re-enables tests for the Docker JDBC data source. These tests were reverted in #4872 due to transitive dependency conflicts introduced by the `docker-client` library. This patch should avoid those problems by using a version of `docker-client` which shades its transitive dependencies and by performing some build-magic to work around problems with that shaded JAR.

In addition, I significantly refactored the tests to simplify the setup and teardown code and to fix several Docker networking issues which caused problems when running in `boot2docker`.

Closes #8101.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Although the tests should automatically pull any missing Docker images, I went ahead and manually pulled them on all of the AMP Jenkins workers. This uncovered an issue where two of the workers' Docker filesystems had entered a safe mode that prevented image pulling, so I logged in and fixed that.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        .image(db.imageName)
+        .networkDisabled(false)
+        .env(db.env.map { case (k, v) => s"$k=$v" }.toSeq.asJava)
+        .exposedPorts(s"${db.jdbcPort}/tcp")
+        .build()
+      containerId = docker.createContainer(config).id
+      // Configure networking (necessary for boot2docker / Docker Machine)
+      val externalPort: Int = {
+        val sock = new ServerSocket(0)
+        val port = sock.getLocalPort
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/DockerJDBCIntegrationSuite.scala]: I always forget your can do this...

**Comment by marmbrus** [comment]
For posterity, is there any easy way to do the parallel pull if we upgrade one of the images or add a new one?

**Comment by JoshRosen** [comment]
Here's the command that I used:

```
pssh -O StrictHostKeyChecking=no -h /root/jenkins_workers.txt -i -x '-t -t' "sudo docker pull mysql:5.7.9"
```

**Comment by marmbrus** [comment]
Tested this locally and other than needing to set `DOCKER_IP` everything worked great.  Talked ot @JoshRosen offline and he's going to try and automate this.

File: sql/core/pom.xml [hunk 2]
```diff
@@ -111,6 +111,19 @@
       <artifactId>mockito-core</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>com.spotify</groupId>
+      <artifactId>docker-client</artifactId>
+      <classifier>shaded</classifier>
+      <version>2.7.7</version>
```
Comment (JoshRosen) [inline sql/core/pom.xml]: Woah, this is an ancient version of `docker-client`. I'll try using a newer release to see if that magically fixes the ASM issues.

**Comment by JoshRosen** [comment]
I think that the test failures that occurred earlier were due to an older incompatible version of ASM being pulled in via Jersey. From a binary compatibility point of view, it should be okay to upgrade from Jersey 1.9 to 1.19: 

- [japi-compliance-checker](https://github.com/lvc/japi-compliance-checker) reports no source or binary compatibility problems for `jersey-server` and `jersey-json`.
- Jersey 1.10's major breaking change was moving the servlet-related classes into a separate `jersey-servlet` artifact, so I added a dependency on that class: https://blogs.oracle.com/japod/entry/jersey_1_10_is_released
- In Jersey 1.19, `jersey-core` removed the `JAX-RS 1.1` classes (see http://blog.dejavu.sk/2015/02/13/jersey-1-19-is-released/), which results in the JAPI checker labeling it as 100% incompatible, but the `javax.ws.rs:jsr311-api:1.1.1` transitive dependency should automatically pull those classes in.

Older Jersey release notes can be found at https://github.com/jersey/jersey-1.x/wiki/Release-Notes.

I'm actually a bit suspicious that the JAPI checker didn't complain about the removal of the servlet classes from `jersey-server`...

What gives me pause, though, is the fact that I found an old Hadoop JIRA discussing upgrading to 1.19 as part of a Java 8 compatibility effort, where it appears that they ran into a number of problems due to incompatible _behavioral_ changes: https://issues.apache.org/jira/browse/HADOOP-9613?focusedCommentId=14972906&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14972906

The ASM conflict stems from the fact that the Spotify Docker client depends on `jnr-unixsocket`, which, in turn, depends on `jnr-ffi`, which uses ASM 5, which is incompatible with the ASM 3 version required by Jersey. This conflict is slightly hard to spot because the Maven artifact names changed between these releases, so Maven doesn't detect the conflict.

I suppose one option would be to just ditch the Spotify Docker client in favor of scripting over the Docker command-line or JSON APIs (or to switch to another Docker client library). Another approach would be to try to exclude the ASM 3.x dependency in tests.

**Comment by JoshRosen** [comment]
Ha, it turns out that `docker-java` isn't a viable option because it also has Jersey conflicts: https://github.com/docker-java/docker-java

**Comment by JoshRosen** [comment]
@marmbrus, I think the reason that we didn't run into this problem is because boot2docker on the Mac doesn't connect using unix sockets: https://groups.google.com/forum/#!topic/docker-user/vO0aPRcwyxU

**Comment by JoshRosen** [comment]
Even though we can't bump Spark's own Jersey version without having to worry about Hadoop conflicts, we might be able to bump the version _just_ in the SQL tests.

**Comment by JoshRosen** [comment]
So I tried out the Jersey upgrade (which isn't a realistic option for all of Spark but might be okay if isolated to a Docker-only test module) and it looks like the tests are now failing with new and exciting errors instead of ASM ones:

```
sbt.ForkMain$ForkError: Request error: POST unix://localhost:80/containers/5b239c9eb6e960828f49ab8030f061ce4bd41a7e0b1bca487357dc97e4c4796a/start: 400
	at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1133)
	at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1104)
	at com.spotify.docker.client.DefaultDockerClient.startContainer(DefaultDockerClient.java:400)
	at org.apache.spark.sql.jdbc.DockerJDBCIntegrationSuite.beforeAll(DockerJDBCIntegrationSuite.scala:115)
	at org.scalatest.BeforeAndAfterAll$class.beforeAll(BeforeAndAfterAll.scala:187)
	at org.apache.spark.sql.jdbc.DockerJDBCIntegrationSuite.beforeAll(DockerJDBCIntegrationSuite.scala:60)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:253)
	at org.apache.spark.sql.jdbc.DockerJDBCIntegrationSuite.run(DockerJDBCIntegrationSuite.scala:60)
	at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:462)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:671)
	at sbt.ForkMain$Run$2.call(ForkMain.java:294)
	at sbt.ForkMain$Run$2.call(ForkMain.java:284)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: sbt.ForkMain$ForkError: HTTP 400 Bad Request
	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:947)
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:795)
	at org.glassfish.jersey.client.JerseyInvocation.access$500(JerseyInvocation.java:91)
	at org.glassfish.jersey.client.JerseyInvocation$5.completed(JerseyInvocation.java:756)
	at org.glassfish.jersey.client.ClientRuntime.processResponse(ClientRuntime.java:189)
	at org.glassfish.jersey.client.ClientRuntime.access$300(ClientRuntime.java:74)
	at org.glassfish.jersey.client.ClientRuntime$1.run(ClientRuntime.java:171)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:320)
	at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:201)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
```

**Comment by JoshRosen** [comment]
I submitted a PR to `docker-client` to try to relocate the ASM classes in the shaded JAR: https://github.com/spotify/docker-client/pull/272. In the meantime, I'll start working on extracting these docker tests to a separate subproject, since I think that's something we should do anyways.

**Comment by JoshRosen** [comment]
It looks like the error when launching the docker conatiner stemmed from a bad IP mapping:

```
time="2015-11-08T17:11:22-08:00" level="error" msg="HTTP Error: statusCode=400 Cannot start container 206b145a60400a8153549bdf10dc79a6298d2e9e72a65ab2ce2602e3c5d54bd0: Bad parameter: invalid host ip localhost"
```

**Comment by JoshRosen** [comment]
Something really weird is going on here.

Following the advice at http://stackoverflow.com/a/21876380/590203, I added the following line to `DockerJDBCIntegrationSuite` in order to figure out where `BasicLineFormatter` was being loaded from:

```
println(this.getClass.getClassLoader.getResource("org/apache/http/message/BasicLineFormatter.class"))
```

When I run this locally with

```
 sbt/sbt -Pyarn -Phadoop-2.3 -Dhadoop.version=2.3.0 -Pkinesis-asl -Phive-thriftserver -Phive
```

I get

```
jar:file:/Users/joshrosen/Documents/spark/lib_managed/jars/httpcore-4.2.4.jar!/org/apache/http/message/BasicLineFormatter.class
```

This doesn't make sense, though: `show test:dependencyClasspath` does not include that JAR, so I'm not quite sure how it's winding up on the classpath. If I toss in a

```
  this.getClass.getClassLoader.asInstanceOf[URLClassLoader].getURLs.foreach(println)
```

I still don't see where that JAR is coming from.

**Comment by JoshRosen** [comment]
In fact, `httpcore-4.2.4` doesn't even show up in the `-classpath` argument passed to `ForkMain`.

I used some of the handy classpath-debugging tools in [JHades](https://github.com/jhades/jhades) to dump more information and it looks like `httpcore-4.2.4` is included in the Tachyon JARs' `MANIFEST.MF` files. Here's the manfest from `tachyon-client`:

```
Manifest-Version: 1.0
Built-By: DreamSweet
Build-Jdk: 1.7.0_75
Class-Path: guava-14.0.1.jar commons-lang-2.4.jar commons-io-2.4.jar c
 ommons-lang3-3.0.jar curator-framework-2.1.0-incubating.jar zookeeper
 -3.4.5.jar slf4j-log4j12-1.7.2.jar log4j-1.2.15.jar mail-1.4.jar acti
 vation-1.1.jar jline-0.9.94.jar netty-3.2.2.Final.jar curator-client-
 2.1.0-incubating.jar curator-recipes-2.1.0-incubating.jar slf4j-api-1
 .7.2.jar hadoop-client-1.0.4.jar hadoop-core-1.0.4.jar xmlenc-0.52.ja
 r commons-codec-1.4.jar commons-math-2.1.jar commons-configuration-1.
 6.jar commons-collections-3.2.1.jar commons-digester-1.8.jar commons-
 beanutils-1.7.0.jar commons-beanutils-core-1.8.0.jar commons-net-1.4.
 1.jar commons-el-1.0.jar hsqldb-1.8.0.10.jar oro-2.0.8.jar jackson-ma
 pper-asl-1.0.1.jar jackson-core-asl-1.0.1.jar tachyon-common-0.8.1.ja
 r netty-all-4.0.28.Final.jar libthrift-0.9.2.jar httpclient-4.2.5.jar
  httpcore-4.2.4.jar tachyon-client-unshaded-0.8.1.jar tachyon-underfs
 -hdfs-0.8.1.jar tachyon-underfs-glusterfs-0.8.1.jar glusterfs-hadoop-
 2.3.13.jar hadoop-yarn-server-nodemanager-2.3.0.jar hadoop-yarn-commo
 n-2.3.0.jar hadoop-yarn-api-2.3.0.jar jaxb-api-2.2.2.jar stax-api-1.0
 -2.jar jettison-1.1.jar servlet-api-2.5.jar jersey-core-1.9.jar jerse
 y-client-1.9.jar jetty-util-6.1.26.jar hadoop-annotations-2.3.0.jar g
 uice-servlet-3.0.jar protobuf-java-2.5.0.jar guice-3.0.jar javax.inje
 ct-1.jar aopalliance-1.0.jar jersey-json-1.9.jar jaxb-impl-2.2.3-1.ja
 r jackson-jaxrs-1.8.3.jar jackson-xc-1.8.3.jar jersey-guice-1.9.jar h
 adoop-yarn-server-common-2.3.0.jar hadoop-common-2.3.0.jar commons-cl
 i-1.2.jar commons-math3-3.1.1.jar jetty-6.1.26.jar jersey-server-1.9.
 jar asm-3.1.jar jasper-compiler-5.5.23.jar jasper-runtime-5.5.23.jar 
 jsp-api-2.1.jar avro-1.7.4.jar paranamer-2.3.jar snappy-java-1.0.4.1.
 jar hadoop-auth-2.3.0.jar jsch-0.1.42.jar jsr305-1.3.9.jar commons-co
 mpress-1.4.1.jar xz-1.0.jar tachyon-underfs-s3-0.8.1.jar jets3t-0.8.1
 .jar commons-logging-1.1.1.jar commons-httpclient-3.1.jar java-xmlbui
 lder-0.4.jar tachyon-underfs-local-0.8.1.jar
Created-By: Apache Maven 3.3.3
Archiver-Version: Plexus Archiver
```

According to the [Java 7 JAR specification](https://docs.oracle.com/javase/7/docs/technotes/guides/jar/jar.html#Manifest_Specification) (emphasis mine):

> Class-Path: The value of this attribute specifies the **relative** URLs of the extensions or libraries that this application or extension needs. URLs are separated by one or more spaces. The application or extension class loader uses the value of this attribute to construct its internal search path.

Assuming that "relative" means "relative to the JAR containing the manifest", I think that what's happening is that this is causing the `httpcore-4.2.4` JAR from `lib_managed` to be added to the classpath, which explains why testing this patch after rebasing on top of #9575 (which minimizes our use of `lib_managed/jars` to only include the Datanucleus JARs) seemed to fix things.

What a giant mess. It looks like the `<addClassPath>` manifest configuration ended up being added to Tachyon's root POM as part of https://github.com/amplab/tachyon/pull/939.

**Comment by pwendell** [comment]
I only reviewed the build changes, but they look good to me.

**Comment by JoshRosen** [comment]
I found this Manifest class-path issue to be extremely puzzling, but it looks like other people have hit it as well: https://dzone.com/articles/jar-manifest-class-path-is-not-for-java-applicatio

Quote:

> A JAR with a Class-Path manifest header will make those classpath entries available to the Java classloader regardless of whether that JAR is executed with java -jar ... or is simply placed on the classpath of a larger Java application.

This puzzling behavior is described a bit more explicitly in [Java 8's JAR specification](https://docs.oracle.com/javase/8/docs/technotes/guides/jar/jar.html#classpath).

**Comment by JoshRosen** [comment]
Created https://tachyon.atlassian.net/browse/TACHYON-1254 for this; this will be worked around in Spark via my `lib_managed` patch, but I still feel that it should be fixed in Tachyon as well so that it doesn't bite anyone in the future.

**Comment by JoshRosen** [comment]
And just to be 100%-sure that it was the Tachyon manifests's fault, I used the following command to delete them and verified that doing so fixed the tests:

```
cd $SPARK_HOME
ls -d -1 lib_managed/jars/* | grep tachyon | xargs -n1 -I{} zip {} --delete META-INF/MANIFEST.MF
```

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Looks like a flaky streaming test?

Jenkins, retest this please.

**Comment by rxin** [comment]
I'm merging this. Thanks.

**Comment by tedyu** [comment]
From https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-Maven-with-YARN/HADOOP_PROFILE=hadoop-2.4,label=spark-test/4095/consoleFull :
```
MySQLIntegrationSuite:
*** RUN ABORTED ***
  java.lang.NoSuchMethodError: org.apache.http.impl.client.HttpClientBuilder.setConnectionManagerShared(Z)Lorg/apache/http/impl/client/HttpClientBuilder;
  at org.glassfish.jersey.apache.connector.ApacheConnector.<init>(ApacheConnector.java:240)
  at org.glassfish.jersey.apache.connector.ApacheConnectorProvider.getConnector(ApacheConnectorProvider.java:115)
  at org.glassfish.jersey.client.ClientConfig$State.initRuntime(ClientConfig.java:418)
  at org.glassfish.jersey.client.ClientConfig$State.access$000(ClientConfig.java:88)
  at org.glassfish.jersey.client.ClientConfig$State$3.get(ClientConfig.java:120)
  at org.glassfish.jersey.client.ClientConfig$State$3.get(ClientConfig.java:117)
  at org.glassfish.jersey.internal.util.collection.Values$LazyValueImpl.get(Values.java:340)
  at org.glassfish.jersey.client.ClientConfig.getRuntime(ClientConfig.java:726)
  at org.glassfish.jersey.client.ClientRequest.getConfiguration(ClientRequest.java:285)
  at org.glassfish.jersey.client.JerseyInvocation.validateHttpMethodAndEntity(JerseyInvocation.java:126)
```

**Comment by markhamstra** [comment]
Similar failure with both `PostgresIntegrationSuite` and `MySQLIntegrationSuite` from `dev/run-tests` on OSX 10.11.1:
```
[info] Exception encountered when attempting to run a suite with class name: org.apache.spark.sql.jdbc.PostgresIntegrationSuite *** ABORTED *** (3 seconds, 40 milliseconds)
[info]   com.spotify.docker.client.DockerException: java.util.concurrent.ExecutionException: com.spotify.docker.client.shaded.javax.ws.rs.ProcessingException: org.apache.http.conn.HttpHostConnectException: Connect to localhost:2375 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1, localhost/fe80:0:0:0:0:0:0:1%1] failed: Connection refused
[info]   at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:1141)
[info]   at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:1082)
[info]   at com.spotify.docker.client.DefaultDockerClient.ping(DefaultDockerClient.java:281)
[info]   at org.apache.spark.sql.jdbc.DockerJDBCIntegrationSuite.beforeAll(DockerJDBCIntegrationSuite.scala:76)
[info]   at org.scalatest.BeforeAndAfterAll$class.beforeAll(BeforeAndAfterAll.scala:187)
...
```

**Comment by JoshRosen** [comment]
Maybe the docker daemon didn't get started / restarted on a subset of the machines?

**Comment by tedyu** [comment]
Should MySQLIntegrationSuite be disabled since the test failure is reproducible ?

**Comment by JoshRosen** [comment]
If you don't have Docker set up locally, you can use the `DockerTest` test tag to exclude those tests.

**Comment by lvc** [comment]
@JoshRosen

The backward compatibility reports for Jersey* libraries are now available in the _API Tracker_ project:

https://abi-laboratory.pro/java/tracker/timeline/jersey-core/
https://abi-laboratory.pro/java/tracker/timeline/jersey-server/

The JAPICC tool detected some servlet related changes in jersey-core 1.10:
https://abi-laboratory.pro/java/tracker/compat_report/jersey-server/1.9.1/1.10/2ae0a/bin_compat_report.html

**Comment by nonsleepr** [comment]
Adding following for better visibility.

The problem while using [`com.whisk:docker-testkit-impl-spotify`](https://github.com/whisklabs/docker-it-scala) along with Spark:
```log
com.spotify.docker.client.exceptions.DockerException: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.IncompatibleClassChangeError: Found interface org.objectweb.asm.ClassVisitor, but class was expected
  at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:2657)
  at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2588)
  at com.spotify.docker.client.DefaultDockerClient.info(DefaultDockerClient.java:522)
  ... 40 elided
Caused by: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.IncompatibleClassChangeError: Found interface org.objectweb.asm.ClassVisitor, but class was expected
  at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299)
  at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:286)
  at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
  at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2586)
  ... 41 more
Caused by: javax.ws.rs.ProcessingException: java.lang.IncompatibleClassChangeError: Found interface org.objectweb.asm.ClassVisitor, but class was expected
  at org.glassfish.jersey.client.ClientRuntime.processFailure(ClientRuntime.java:202)
  at org.glassfish.jersey.client.ClientRuntime.access$400(ClientRuntime.java:79)
  at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:182)
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
  at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
  at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
  at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:340)
  at org.glassfish.jersey.client.ClientRuntime$3.run(ClientRuntime.java:210)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  ... 1 more
Caused by: java.lang.IncompatibleClassChangeError: Found interface org.objectweb.asm.ClassVisitor, but class was expected
  at jnr.ffi.provider.jffi.AsmLibraryLoader.generateInterfaceImpl(AsmLibraryLoader.java:74)
  at jnr.ffi.provider.jffi.AsmLibraryLoader.loadLibrary(AsmLibraryLoader.java:59)
  at jnr.ffi.provider.jffi.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:43)
  at jnr.ffi.LibraryLoader.load(LibraryLoader.java:287)
  at jnr.unixsocket.Native.<clinit>(Native.java:76)
  at jnr.unixsocket.UnixSocketChannel.<init>(UnixSocketChannel.java:68)
  at jnr.unixsocket.UnixSocketChannel.open(UnixSocketChannel.java:49)
  at com.spotify.docker.client.ApacheUnixSocket.<init>(ApacheUnixSocket.java:59)
  at com.spotify.docker.client.UnixConnectionSocketFactory.createSocket(UnixConnectionSocketFactory.java:67)
  at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:118)
  at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
  at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
  at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
  at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
  at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
  at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
  at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
  at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
  at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:435)
  at org.glassfish.jersey.apache.connector.ApacheConnector$1.run(ApacheConnector.java:491)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  at jersey.repackaged.com.google.common.util.concurrent.MoreExecutors$DirectExecutorService.execute(MoreExecutors.java:299)
  at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:112)
  at jersey.repackaged.com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:50)
  at jersey.repackaged.com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:37)
  at org.glassfish.jersey.apache.connector.ApacheConnector.apply(ApacheConnector.java:487)
  at org.glassfish.jersey.client.ClientRuntime$2.run(ClientRuntime.java:178)
  ... 12 more
```

The solution is to use `shaded` package:
```scala
libraryDependencies += "com.spotify" % "docker-client" % "8.11.3" classifier "shaded"
````

**Pull request closed without merging**